sdk/dotnet/Pulumi.Automation/Events/EventLogWatcher.cs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2016-2021, Pulumi Corporation
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Pulumi.Automation.Serialization;
+
+namespace Pulumi.Automation.Events
+{
+    internal class EventLogWatcher : IAsyncDisposable
```

Comment:
OK this is pretty cool. I did not quite understand cancellation so I tried to write my own version and test it out and eventually ended up with something extremely similar to this so it may not be worth to change much here.

The principal difference where I landed was handling exceptions, might be worth a consideration. I err on the side of never ignoring exceptions either from polling code or from user callbacks, and instead terminate the polling task with the first exception and propagate it to the code that is calling EventLogWatcher, so it would fail also. It appears from IAsyncDisposable interface that it might not be appropriate to throw from DisposeAsync, so I ended up with a simple IDisposable but the cleanup is a separate call:

```
async Task Stop();
``` 

And then we basically propagate any exception that might have been raised as a result of this Stop task.

Reply:
I don't think we need to propagate exceptions from cleanup - since it is us attempting to delete a temp folder as a courtesy, that the user's OS should be cleaning up on its own anyways. We have precedence for ignoring an error here in LocalWorkspace disposal as well.

But I agree we should propagating the exceptions up everywhere else.

Reply:
https://github.com/vipentti/pulumi/pull/2/files the version I ended up with for reference.

Reply:
gotcha. Yea it does make sense to separate out the `Stop` function from the dispose if we want to propagate errors from it. Where directory cleanup would remain in the `Dispose` action